### PR TITLE
Add blood vessel junction

### PR DESCRIPTION
**Pull request by richterjakob**

### CODE REVIEW

File: src/model/bloodvesseljunction.hpp [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+  // Set number of equations of a junction block based on number of
+  // inlets/outlets. Must be set before calling parent constructor
+  num_outlets = this->outlet_nodes.size();
+  std::list<std::string> internal_var_names;
+  for (size_t i = 0; i < num_outlets; i++) {
+    internal_var_names.push_back("flow_" + std::to_string(i));
+  }
+  Block<T>::setup_dofs_(dofhandler, 1, internal_var_names);
+  for (size_t i = 0; i < num_outlets; i++) {
+    blood_vessels.push_back(new BloodVessel<T>(
```
Comment (mrp089) [inline src/model/bloodvesseljunction.hpp:212]: Perfect! I'm a big fan of your implementation reusing the `BloodVessel` object!

File: src/model/bloodvesseljunction.hpp [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+ * The governing equations are mainly defined by the blood vessel elements
+ * (see MODEL::BloodVessel for more information). For the sake of brevity,
+ * the blood vessel contributions are not repeated here. In addition to the
+ * individual blood vessel contributions, the following equation ensures
+ * mass conservation between the inlets of the blood vessel elements:
+ *
+ * \f[
+ * Q_{in}=\sum_{i}^{n_{outlets}} Q_{in, i}
+ * \f]
+ *
```
Comment (mrp089) [inline src/model/bloodvesseljunction.hpp:74]: Well documented!

File: src/model/bloodvesseljunction.hpp [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+template <typename T>
+void BloodVesselJunction<T>::update_constant(ALGEBRA::SparseSystem<T> &system) {
+  for (auto bv : blood_vessels) {
+    bv->update_constant(system);
+  }
+  // Mass conservation
+  system.F.coeffRef(this->global_eqn_ids[0], this->global_var_ids[1]) = 1.0;
+  for (size_t i = 0; i < num_outlets; i++) {
+    system.F.coeffRef(this->global_eqn_ids[0],
+                      this->global_var_ids.end()[-(i + 1)]) = -1.0;
```
Comment (mrp089) [inline src/model/bloodvesseljunction.hpp:237]: @richterjakob, you have to judge if it's worth it: Would it make sense to move mass conservation into a junction base class? Physically, we would expect this from every junction, regardless of how we model the pressure change.

From this, we could derive several specific junctions (that we already have): "normal" junction that preserves static pressure, resistive junction, BloodVessel junction.

This structure could make it slightly easier and less redundant (but more confusing) to add new junctions, like preserving total pressure or advanced models like @natalia-rubio is working on.

File: src/model/bloodvesseljunction.hpp [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+ * \draw (1,0) to [short, *-] (2.5,-0.75);
+ * \draw (2.5,0.75) node[anchor=south]{} to [generic, l_=$BV_{1}$, -*]
+ * (4.5,0.75); \draw (2.4,0.75) node[anchor=south]{$Q_{in,1}$}; \draw (4.6,0.75)
+ * node[anchor=south] {$P_{out,1}$}; \draw (2.5,-0.75) node[anchor=south]{} to
+ * [generic, l^=$BV_{2}$, -*] (4.5,-0.75); \draw (2.4,-0.75)
+ * node[anchor=north]{$Q_{in,2}$}; \draw (4.6,-0.75) node[anchor=north]
+ * {$P_{out,2}$}; \draw [-latex] (4.7,0.75) -- (5.5,0.75) node[right]
+ * {$Q_{out,1}$}; \draw [-latex] (4.7,-0.75) -- (5.5,-0.75) node[right]
+ * {$Q_{out,2}$}; \end{circuitikz} \f]
+ *
```
Comment (mrp089) [inline src/model/bloodvesseljunction.hpp:62]: Technical question: Can I view these somewhere while the pull request is still open?

**Review by mrp089 — CHANGES_REQUESTED** [review]
@richterjakob, well done! Implementation-wise perfect and well documented!

Before we merge, I have two small requests.

To make @ktbolt's day, can you open a super brief issue on what the `BloodVesselJunction` is and why we want it? Then tag the issue here so we can find it later (will be closed once this is merged).

Can you add a brief test case that uses the `BloodVesselJunction`?  It doesn't need to be super complicated, something simple where we have an analytic solution (steady-state flow and resistance-only) like in the existing junction test cases would already be helpful.

Thanks a lot! I'm super excited you have already implemented this!

File: src/model/bloodvesseljunction.hpp (hunk 3 above)
Comment (richterjakob) [reply src/model/bloodvesseljunction.hpp:237]: The mass conservation here is not the same as in the normal junction class. For the normal junction, mass is conserved between inlets and outlets of the junction. This mass conservation here is a little hacky way to enable reusing the BloodVessel class. It ensures that the flow into the junction is correctly distributed to the **inlets** of the internal BloodVessel elements. I know this is a bit confusing and I didn't find a way to describe it better. Let me know, what you think. Should I add more documentation on that?

File: src/model/bloodvesseljunction.hpp (hunk 4 above)
Comment (richterjakob) [reply src/model/bloodvesseljunction.hpp:62]: Yes, the html documentation is automatically built on GitHub. You can always go to the [last run of the Documentation routine](https://github.com/StanfordCBCL/svZeroDPlus/actions/runs/3525740588) and download the `documentation` artifact. You can then open `cpp`->`index.html` in the downloaded folder to view the html documentation locally.

**Comment by richterjakob** [comment]
I created issue #19 corresponding to this pull request 👍

**Comment by richterjakob** [comment]
I added a test case in the latest commit. Let me know if you have more suggestions.

File: src/model/bloodvesseljunction.hpp (hunk 3 above)
Comment (mrp089) [reply src/model/bloodvesseljunction.hpp:237]: Ah, I see your indexing is different. I think that's fine. I don't think it makes sense to change all indexing for the sake of inheritance.

**Review by mrp089 — APPROVED** [review]

**Pull request merged**